cudax/include/cuda/experimental/__event/event_ref.cuh

```diff
@@ -0,0 +1,132 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX_EVENT_REF_DETAIL_H
+#define _CUDAX_EVENT_REF_DETAIL_H
+
+#include <cuda_runtime_api.h>
+// cuda_runtime_api needs to come first
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cuda/api_wrapper.h>
+#include <cuda/std/__exception/cuda_error.h>
+#include <cuda/std/cassert>
+#include <cuda/std/cstddef>
+#include <cuda/std/utility>
+#include <cuda/stream_ref>
+
+namespace cuda::experimental
+{
+class event;
+class timed_event;
+
+//! @brief An non-owning wrapper for an untimed `cudaEvent_t`.
+class event_ref
+{
+private:
+  friend class event;
+  friend class timed_event;
+
+  ::cudaEvent_t __event_{0};
+
+public:
+  using value_type = ::cudaEvent_t;
+
+  //! @brief Construct a new `event_ref` that does refer to an event.
+  event_ref() = default;
```

Comment:
I think event_ref should not have this constructor, at least not public

Reply:
:+1:

Reply:
i'm curious, why? and (directed at @miscco) does your reasoning also apply to `stream_ref`?

Reply:
I would say each of the *_ref objects like a real reference should point to a valid object. For stream_ref right now the argument less constructor refers to the NULL stream, there are no valid "default" events like that in CUDA. I would even argue that we should remove it from stream_ref, because there will be no NULL stream in the C++ runtime, but I don't know if we can do it for an already released API

Reply:
> each of the *_ref objects like a real reference should point to a valid object

but that's not possible if we want efficient moves of an owning event type. consider:

```c++
event e(event::flags::none); // calls cudaEventCreateWithFlags
event e2 = std::move(e);     // e becomes null
event_ref r = e;             // r is an null event_ref
```

Reply:
I think its fine with moved-out of objects, it's on the user to use move semantic correctly.
If we want this for arrays or members, I would prefer for this to be some sort of opt-in like cuda::event(cudaEvent_t{}) or maybe cuda::event(cuda::uninit_t) instead of the argument less constructor, which is easy to be used accidentally.

Reply:
@pciolkosz already said everything about it. 

I am wary of the user being able to accidentally create something meaningless. And accidentally is a strong word when it comes to default construction. 

I would much prefer if we could keep those functions internal and not expose it to the user.